compiler/transforms/dispatch_kernels.py

```diff
@@ -0,0 +1,133 @@
+from xdsl.dialects import builtin, linalg
+from xdsl.dialects.memref import MemRefType
+from xdsl.ir import MLContext
+from xdsl.ir.affine import AffineDimExpr, AffineMap
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+from compiler.util.kernel_type import KernelType
+
+
+class DispatchElementwiseMult(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: linalg.Generic, rewriter: PatternRewriter):
+        """Add a library call to a linalg generic that implements an
+        elementwise multiplication. This is done on linalg.generics as
+        named linalg ops do not support library calls. This makes the task
+        of detecting an elementwise multiplication somewhat harder,
+        but this can be more structured in future work."""
+
+        ## conditions for library call:
+        # (1) kernel type must be mul
+        # (2) data type must be 1D integer memref
+        # (3) dataflow must be all parallel
+
+        if op.library_call is not None:
+            return
+
+        if KernelType.get_kernel(op) != KernelType.MUL:
+            return
+
+        for inp in op.inputs:
+            if not isinstance(inp.type, MemRefType):
+                return
+
+            if len(inp.type.get_shape()) > 1:
+                return
+
+            if not isinstance(inp.type.get_element_type(), builtin.IntegerType):
+                return
+
+        if len(op.iterator_types) != 1:
+            return
+
+        if op.iterator_types.data[0].data is not linalg.IteratorType.PARALLEL:
+            return
```

Comment:
Not sure if we can reorder the for loop out of this, but again I think we can fuse some of this stuff together?
Let me know if you don't like this style!
```suggestion
        if any( 
            op.library_call is not None,
            KernelType.get_kernel(op) != KernelType.MUL,
            ...
         ):
            return
        for inp in op.inputs:
             if any(
                         ...
                    ):
                    return
       # Move these up?
       # if len(op.iterator_types) != 1:
       #     return

       # if op.iterator_types.data[0].data is not linalg.IteratorType.PARALLEL:
       #     return
```

Reply:
On second thought, I think applying my suggestions messes up performance :( :
https://stackoverflow.com/a/53600865/18472362